core/src/main/java/org/apache/iceberg/ManifestReader.java

```diff
@@ -369,4 +390,28 @@ static List<String> withStatsColumns(Collection<String> columns) {
       return projectColumns;
     }
   }
+
+  private static <F extends ContentFile<F>> Function<ManifestEntry<F>, ManifestEntry<F>> idAssigner(
```

Comment:
Keeping state in `InheritableMetadata` doesn't work because it is reused each time the reader produces an iterator. That causes incorrect row ID assignment (caught by the tests I'm working on). Instead, I've introduced this assigner function that is called where `InheritableMetadata` is used. `InheritableMetadata` is used for constants, this is used for state-based ID assignment.